File: December 2022/03 December 2022/C++/Aggressive Cows.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
public:
    bool ispossible(int n, int k, vector<int> &stalls, int dist) {

        int coord = stalls[0];

        int cnt = 1;

        for (int i = 1; i < n; i++) {
            if (stalls[i] - coord >= dist) {
                cnt++;
                coord = stalls[i];
            }

            if (cnt == k) return true;
        }
        return false;
    }

    int solve(int n, int k, vector<int> &stalls){
        
        sort(stalls.begin(),stalls.end());
        int low=1;
        int high=stalls[n-1]-stalls[0];
       
        while(low<=high)
        {
            int mid=(low+high)/2;
            if(ispossible(n,k,stalls,mid)) low=mid+1;
            else high=mid-1;
        }
        return high;
    }
};

//{ Driver Code Starts.

int main() {
    int t = 1;
    cin >> t;

    // freopen ("output_gfg.txt", "w", stdout);

    while (t--) {
        // Input

        int n, k;
        cin >> n >> k;

        vector<int> stalls(n);
        for (int i = 0; i < n; i++) {
            cin >> stalls[i];
        }
        // char ch;
        // cin >> ch;

        Solution obj;
        cout << obj.solve(n, k, stalls) << endl;

        // cout << "~\n";
    }
    // fclose(stdout);

    return 0;
}
// } Driver Code Ends

File: December 2022/03 December 2022/Python/Aggressive Cows.py
#User function Template for python3

class Solution:
    def solve(self, n, k, stalls):
        stalls.sort()
        
        def count(d):
            ans, curr = 1, stalls[0]
            for i in range(1, n):
                if stalls[i] - curr >= d:
                    ans += 1
                    curr = stalls[i]
            return ans
        
        l, r = 0, stalls[-1] - stalls[0]
        while l < r:
            mid = r - (r - l) // 2
            if count(mid) >= k:
                l = mid
            else:
                r = mid - 1
        return l

#{ 
 # Driver Code Starts
#Initial Template for Python 3

def main():

    T = int(input())

    while(T > 0):
        n, k = list(map(int, input().split()))
        stalls = list(map(int, input().split()))
        ob = Solution()
        res = ob.solve(n, k, stalls)
        print(res)

        T -= 1


if __name__ == "__main__":
    main()
# } Driver Code Ends